docs/source/developers/TESTING.md

```diff
@@ -0,0 +1,168 @@
+
+```sh
+# run the full test suite with coverage
+pytest --cov=napari
+# instead of coverage in the console, get a nice browser-based cov-report
+pytest --cov=napari --cov-report=html
+open htmlcov/index.html  # look at the report
+# run a subset of tests with coverage
+pytest --cov=napari.layers.shapes --cov-report=html napari/layers/shapes
+open htmlcov/index.html  # look at the report
+```
+
+## Writing Tests
+
+Writing tests for new code is a critical part of keeping napari maintainable as
+it grows. Tests are written in files whose names
+begin with `test_*` and which are contained in one of the `_tests` directories.
+
+There are a couple things to keep in mind when writing a test where a `Qt` event
+loop or a `napari.Viewer` is required.  The important thing is that any widgets
+you create during testing are cleaned up at the end of each test:
+
+1. If you need a `QApplication` to be running for your test, you can use the
+   [`qtbot`](https://pytest-qt.readthedocs.io/en/latest/reference.html#pytestqt.qtbot.QtBot) fixture from `pytest-qt`
+
+    > note: fixtures in pytest can be a little mysterious, since it's not always
+    > clear where they are coming from.  In this case, using a pytest-qt fixture
+    > looks like this:
+
+    ```python
+    # just by putting `qtbot` in the list of arguments
+    # pytest-qt will start up an event loop for you
+    def test_something(qtbot):
+        ...
+    ```
+
+   `qtbot` provides a convenient
+   [`addWidget`](https://pytest-qt.readthedocs.io/en/latest/reference.html#pytestqt.qtbot.QtBot.addWidget)
+   method that will ensure that the widget gets closed at the end of the test.
+   It *also* provides a whole bunch of other
+   convenient methods for interacting with your GUI tests (clicking, waiting
+   signals, etc...).  See the [`qtbot` docs](https://pytest-qt.readthedocs.io/en/latest/reference.html#pytestqt.qtbot.QtBot) for details.
+
+    ```python
+    # the qtbot provides convenience methods like addWidget
+    def test_something_else(qtbot):
+        widget = QWidget()
+        qtbot.addWidget(widget)  # tell qtbot to clean this widget later
+        ...
+    ```
+
+2. When writing a test that requires a `napari.Viewer` object, we provide our
+   own convenient fixture called `viewer_factory` that will take care of
+   creating a viewer and cleaning up at the end of the test.  When using this
+   function, it is **not** necessary to use a `qtbot` fixture, nor should you do
+   any additional cleanup (such as using `qtbot.addWidget` or calling
+   `viewer.close()`) at the end of the test.  Duplicate cleanup may cause an
+   error.  Use the factory as follows:
+
+    ```python
+    # the viewer_factory fixture is defined in napari/conftest.py
+    def test_something_with_a_viewer(viewer_factory):
+        # viewer factory takes any keyword arguments that napari.Viewer() takes
+        view, viewer = viewer_factory()
+        # note, `view` here is just a pointer to viewer.window.qt_viewer
```

Comment:
Incidentally, how necessary is that pointer? Maybe we should just return `viewer` and use the pattern `view = viewer.window.qt_viewer` where needed?

Reply:
very unnecessary!  i think it's vestigial from some early test patterns.  I like the idea of just returning the viewer.  I'll open a PR for that and we can change these docs

Reply:
That PR #1386 looks great to me. How about I merge this PR first and then you can update docs over there?

Reply:
sure